File: frontend/lib/api/callcenter.ts
import { http } from "./http";

export type CallDirection = "inbound" | "outbound";
export type CallOutcome = "completed" | "missed" | "voicemail" | "transferred";
export type AgentStatus = "available" | "on-call" | "away" | "offline";

export interface QueuedCall {
  id: string;
  name: string;
  number: string;
  reason: string;
  waitedSec: number;
}

export interface RecentCall {
  id: string;
  name: string;
  number: string;
  direction: CallDirection;
  outcome: CallOutcome;
  durationSec: number;
  at: string;
  recorded: boolean;
}

export interface ActiveCall {
  id: string;
  name: string;
  number: string;
  direction: CallDirection;
  startedAt: string;
  muted: boolean;
  onHold: boolean;
}

export interface AgentRow {
  id: string;
  name: string;
  initials: string;
  status: AgentStatus;
  callsToday: number;
  isSelf: boolean;
  sipExtension: string | null;
}

export interface CallStats {
  callsToday: number;
  callsDelta: number;
  avgHandleSec: number;
  ahtDelta: number;
  answerRate: number;
  answerDelta: number;
  missed: number;
  missedDelta: number;
}

export interface VolumePoint {
  name: string;
  Inbound: number;
  Outbound: number;
}

export interface CallOverview {
  stats: CallStats;
  queue: QueuedCall[];
  recent: RecentCall[];
  agents: AgentRow[];
  volume: VolumePoint[];
  active: ActiveCall | null;
  presence: AgentStatus;
  ivrCalls: IvrCall[];
}

export interface IvrCall {
  id: string;
  name: string;
  number: string;
  menu: string;
  prompt: string;
}

export interface CallPoll {
  queue: QueuedCall[];
  active: ActiveCall | null;
  presence: AgentStatus;
  ivrCalls: IvrCall[];
}

export type IvrAction =
  | "ring_all"
  | "ring_agent"
  | "submenu"
  | "voicemail"
  | "message"
  | "transfer"
  | "hangup"
  | "repeat";

export interface IvrOption {
  digit: string;
  label: string;
  action: IvrAction;
  target: string;
}

export interface IvrMenu {
  prompt: string;
  options: IvrOption[];
}

export interface IvrFlow {
  is_active: boolean;
  greeting: string;
  invalid_message: string;
  timeout_message: string;
  timeout_seconds: number;
  max_retries: number;
  on_exhausted: "ring_all" | "voicemail" | "hangup";
  menus: Record<string, IvrMenu>;
  hours_enabled: boolean;
  timezone: string;
  hours: Record<string, string[][]>;
  after_hours_action: "ring_all" | "voicemail" | "hangup" | "message";
  after_hours_message: string;
}

export interface IvrSimulateResult {
  transcript: { kind: string; text: string }[];
  resolved: { action: string; target: string; label: string };
}

export interface SoftphoneConfig {
  /** true only when a real SIP trunk is active AND this agent has a line assigned */
  configured: boolean;
  provider: string; // simulated | cloudone | asterisk
  ws_url: string | null;
  domain: string | null;
  extension: string | null;
  password: string | null;
  display_name: string | null;
  turn_url: string | null;
  turn_user: string | null;
  turn_password: string | null;
}

export const callCenterApi = {
  overview: () => http.get<CallOverview>("/call-center/overview"),
  poll: () => http.get<CallPoll>("/call-center/poll"),
  dial: (name: string, number: string) =>
    http.post<CallOverview>("/call-center/dial", { name, number }),
  answer: (id: string) => http.post<CallOverview>(`/call-center/calls/${id}/answer`),
  hangup: (id: string, outcome: CallOutcome = "completed") =>
    http.post<CallOverview>(`/call-center/calls/${id}/hangup`, { outcome }),
  flags: (id: string, flags: { muted?: boolean; on_hold?: boolean }) =>
    http.post<CallOverview>(`/call-center/calls/${id}/flags`, flags),
  dismiss: (id: string) => http.post<CallOverview>(`/call-center/calls/${id}/dismiss`),
  presence: (status: AgentStatus) =>
    http.post<CallOverview>("/call-center/presence", { status }),
  simulateInbound: () => http.post<CallOverview>("/call-center/simulate-inbound"),
  softphoneConfig: () => http.get<SoftphoneConfig>("/call-center/softphone"),
  setAgentSip: (agentId: string, sip_extension: string, sip_password: string) =>
    http.patch<CallOverview>(`/call-center/agents/${agentId}/sip`, { sip_extension, sip_password }),

  getIvr: () => http.get<IvrFlow>("/call-center/ivr"),
  updateIvr: (patch: Partial<IvrFlow>) => http.put<IvrFlow>("/call-center/ivr", patch),
  testIvr: (digits: string[]) =>
    http.post<IvrSimulateResult>("/call-center/ivr/test", { digits }),
  simulateIvrCall: () => http.post<CallOverview>("/call-center/ivr/simulate-call"),
  ivrPress: (callId: string, digit: string) =>
    http.post<CallOverview>(`/call-center/calls/${callId}/ivr-press`, { digits: [digit] }),
};

File: backend/app/schemas/call_center.py
from typing import Any, Dict, List, Optional

from pydantic import BaseModel, Field


class QueuedCall(BaseModel):
    id: str
    name: str
    number: str
    reason: str
    waitedSec: int


class RecentCall(BaseModel):
    id: str
    name: str
    number: str
    direction: str
    outcome: str
    durationSec: int
    at: str
    recorded: bool


class ActiveCall(BaseModel):
    id: str
    name: str
    number: str
    direction: str
    startedAt: str
    ringing: bool = False   # dialled/answered-pending — the UI shows no timer yet
    muted: bool
    onHold: bool


class AgentRow(BaseModel):
    id: str
    name: str
    initials: str
    status: str
    callsToday: int
    isSelf: bool = False
    sipExtension: Optional[str] = None


class CallStats(BaseModel):
    callsToday: int
    callsDelta: float
    avgHandleSec: int
    ahtDelta: float
    answerRate: float
    answerDelta: float
    missed: int
    missedDelta: float


class VolumePoint(BaseModel):
    name: str
    Inbound: int
    Outbound: int


class IvrCall(BaseModel):
    id: str
    name: str
    number: str
    menu: str
    prompt: str = ""


class CallOverview(BaseModel):
    stats: CallStats
    queue: List[QueuedCall]
    recent: List[RecentCall]
    agents: List[AgentRow]
    volume: List[VolumePoint]
    active: Optional[ActiveCall] = None
    presence: str
    ivrCalls: List[IvrCall] = []


class CallPoll(BaseModel):
    queue: List[QueuedCall]
    active: Optional[ActiveCall] = None
    presence: str
    ivrCalls: List[IvrCall] = []


class DialRequest(BaseModel):
    name: str = ""
    number: str = Field(..., min_length=3, max_length=40)


class HangupRequest(BaseModel):
    outcome: str = "completed"


class FlagsRequest(BaseModel):
    muted: Optional[bool] = None
    on_hold: Optional[bool] = None


class PresenceRequest(BaseModel):
    status: str


class SoftphoneEvent(BaseModel):
    # inbound_ring | inbound_answered | outbound_answered | ended | declined
    kind: str = Field(..., max_length=32)
    number: Optional[str] = Field(default=None, max_length=40)
    name: Optional[str] = Field(default=None, max_length=200)


class SoftphoneConfig(BaseModel):
    configured: bool
    provider: str = "simulated"
    ws_url: Optional[str] = None
    domain: Optional[str] = None
    extension: Optional[str] = None
    password: Optional[str] = None
    display_name: Optional[str] = None
    # TURN relay so the browser's audio reaches the PBX even when direct
    # ICE fails (agent behind carrier-grade NAT, etc.)
    turn_url: Optional[str] = None
    turn_user: Optional[str] = None
    turn_password: Optional[str] = None


class AgentSipUpdate(BaseModel):
    sip_extension: Optional[str] = None
    sip_password: Optional[str] = None  # write-only; "" clears


class IvrOption(BaseModel):
    digit: str
    label: str = ""
    action: str
    target: str = ""


class IvrMenu(BaseModel):
    prompt: str = ""
    options: List[IvrOption] = []


class IvrFlowPayload(BaseModel):
    is_active: bool = False
    greeting: str = ""
    invalid_message: str = "Sorry, that isn't a valid option."
    timeout_message: str = "We didn't catch that."
    timeout_seconds: int = 7
    max_retries: int = 2
    on_exhausted: str = "ring_all"
    menus: Dict[str, IvrMenu] = {}
    hours_enabled: bool = False
    timezone: str = "Africa/Nairobi"
    hours: Dict[str, List[List[str]]] = {}
    after_hours_action: str = "voicemail"
    after_hours_message: str = ""


class IvrFlowUpdate(BaseModel):
    is_active: Optional[bool] = None
    greeting: Optional[str] = None
    invalid_message: Optional[str] = None
    timeout_message: Optional[str] = None
    timeout_seconds: Optional[int] = None
    max_retries: Optional[int] = None
    on_exhausted: Optional[str] = None
    menus: Optional[Dict[str, IvrMenu]] = None
    hours_enabled: Optional[bool] = None
    timezone: Optional[str] = None
    hours: Optional[Dict[str, List[List[str]]]] = None
    after_hours_action: Optional[str] = None
    after_hours_message: Optional[str] = None


class IvrSimulateRequest(BaseModel):
    digits: List[str] = Field(default_factory=list, max_length=20)


class IvrTranscriptLine(BaseModel):
    kind: str
    text: str


class IvrResolved(BaseModel):
    action: str
    target: str = ""
    label: str = ""


class IvrSimulateResult(BaseModel):
    transcript: List[IvrTranscriptLine]
    resolved: IvrResolved


class CallResult(BaseModel):
    """Generic mutation response — the fresh poll payload."""
    queue: List[QueuedCall]
    active: Optional[ActiveCall] = None
    presence: str
    recent: Optional[List[RecentCall]] = None
    stats: Optional[CallStats] = None
    agents: Optional[List[AgentRow]] = None
    volume: Optional[List[VolumePoint]] = None
